### PR TITLE
Update Android build setup

### LINKED ISSUE #912: Fails to build with libc++ and clang 18.x
When building with clang and using llvm c++ runtime ( libc++ ), the builds fail with

```
| In file included from /mnt/b/yoe/master/build/tmp/work/riscv64-yoe-linux/vulkan-samples/git/git/framework/platform/platform.cpp:18:
| In file included from /mnt/b/yoe/master/build/tmp/work/riscv64-yoe-linux/vulkan-samples/git/git/framework/platform/platform.h:21:
| In file included from /mnt/b/yoe/master/build/tmp/work/riscv64-yoe-linux/vulkan-samples/git/recipe-sysroot/usr/include/c++/v1/string:62
5:
| In file included from /mnt/b/yoe/master/build/tmp/work/riscv64-yoe-linux/vulkan-samples/git/recipe-sysroot/usr/include/c++/v1/string_vi
ew:213:
| /mnt/b/yoe/master/build/tmp/work/riscv64-yoe-linux/vulkan-samples/git/recipe-sysroot/usr/include/c++/v1/__fwd/string_view.h:22:41: erro
r: 'char_traits<fmt::detail::char8_type>' is deprecated: char_traits<T> for T not equal to char, wchar_t, char8_t, char16_t or char32_t i
s non-standard and is provided for a temporary period. It will be removed in LLVM 19, so please migrate off of it. [-Werror,-Wdeprecated-
declarations]
|    22 | template <class _CharT, class _Traits = char_traits<_CharT> >
|       |                                         ^
| /mnt/b/yoe/master/build/tmp/work/riscv64-yoe-linux/vulkan-samples/git/git/third_party/fmt/include/fmt/core.h:357:26: note: in instantia
tion of default argument for 'basic_string_view<fmt::detail::char8_type>' required here
|   357 | template <typename Char> using std_string_view = std::basic_string_view<Char>;
|       |                          ^~~~~
| /mnt/b/yoe/master/build/tmp/work/riscv64-yoe-linux/vulkan-samples/git/git/third_party/fmt/include/fmt/core.h:448:50: note: in instantia
tion of template type alias 'std_string_view' requested here
|   448 |                                       S, detail::std_string_view<Char>>::value)>
|       |                                                  ^
| /mnt/b/yoe/master/build/tmp/work/riscv64-yoe-linux/vulkan-samples/git/git/third_party/fmt/include/fmt/format.h:781:57: note: in instant
iation of template class 'fmt::basic_string_view<fmt::detail::char8_type>' requested here
|   781 | inline auto compute_width(basic_string_view<char8_type> s) -> size_t {
|       |                                                         ^
| /mnt/b/yoe/master/build/tmp/work/riscv64-yoe-linux/vulkan-samples/git/recipe-sysroot/usr/include/c++/v1/__string/char_traits.h:81:8: no
te: 'char_traits<fmt::detail::char8_type>' has been explicitly marked deprecated here
|    81 | struct _LIBCPP_DEPRECATED_(
|       |        ^
| /mnt/b/yoe/master/build/tmp/work/riscv64-yoe-linux/vulkan-samples/git/recipe-sysroot/usr/include/c++/v1/__config:975:53: note: expanded
 from macro '_LIBCPP_DEPRECATED_'
|   975 | #      define _LIBCPP_DEPRECATED_(m) __attribute__((__deprecated__(m)))
|       |                                                     ^
| 1 error generated.
```

For now the workaround is to use `-Wno-error=deprecated-declarations` in cppflags but it will be a hard error starting llvm-19 so this needs to be fixed to not use char_traits<T>

### LINKED ISSUE #1178: Fails to build for Android with `A problem was found with the configuration of task ':app:mergeDebugJniLibFolders'`
I am trying to build for Android and I get this same error when building from both a macbook and windows desktop.

Android Studio version:

```
Android Studio Koala Feature Drop | 2024.1.2
Build #AI-241.18034.62.2412.12266719, built on August 23, 2024
Runtime version: 17.0.11+0--11852314 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 11.0
GC: G1 Young Generation, G1 Old Generation
Memory: 4096M
Cores: 16
Registry:
  ide.experimental.ui=true
Non-Bundled Plugins:
  Dart (241.18808)
  com.developerphil.adbidea (1.6.17)
  io.flutter (81.0.2)
```

Error:

```
* What went wrong:
A problem was found with the configuration of task ':app:mergeDebugJniLibFolders' (type 'MergeSourceSetFolders').
  - Gradle detected a problem with the following location: 'C:\Users\Soumik\werk\Vulkan-Samples\build\android_gradle\layerLib\jniLibs\android-binaries-1.3.275.0'.
    
    Reason: Task ':app:mergeDebugJniLibFolders' uses this output of task ':app:unzip' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':app:unzip' as an input of ':app:mergeDebugJniLibFolders'.
      2. Declare an explicit dependency on ':app:unzip' from ':app:mergeDebugJniLibFolders' using Task#dependsOn.
      3. Declare an explicit dependency on ':app:unzip' from ':app:mergeDebugJniLibFolders' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.7/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
```

I have tried to add that as a dependency using dependsOn, but then it just says no such task "mergeDebugJniLibFolders" was defined in app.

### LINKED ISSUE #1188: Update android build setup
We need to update our android build setup and make sure it works with recent Android build tools and IDEs. Some recent issues suggest that this is no longer the case and people can't build our samples for Android.

I'll take a look at that.

### LINKED ISSUE #1243: Andriod build failed with ld error:  relocation R_AARCH64_CONDBR19 out of range: 1256256 is not in [-1048576, 1048575]; references __memcpy
My dev environment:
```
Vulkan_Samples revision:  bee3fcfd73ac9f33d7139c1b0c6ff351e5ee259f@main
NDK_ROOT=C:\Users\zieng\AppData\Local\Android\Sdk\ndk\25.2.9519653
ANDROID_HOME=C:\Users\zieng\AppData\Local\Android\Sdk
JAVA_HOME=C:\Program Files\Java\jdk-23
cmake version 3.31.0
Gradle 8.11.1
Kotlin:        2.0.20
Groovy:        3.0.22
Ant:           Apache Ant(TM) version 1.10.14 compiled on August 16 2023
Launcher JVM:  23.0.1 (Oracle Corporation 23.0.1+11-39)
Daemon JVM:    C:\Program Files\Java\jdk-23 (no JDK specified, using current Java home)
OS:            Windows 10 10.0 amd64

java version "23.0.1" 2024-10-15
```
I followed the instruction from `docs\build.adoc`:
```
python .\scripts\generate.py android
cd .\build\android_gradle\
gradle assembleDebug
```
I tried different ndk version but still get the same build failures. I then noticed in `bldsys\cmake\template\gradle\app.build.gradle.in` the ndkVersion is `25.2.9519653`, so I downloaded the exact same version, but still there is the build failure.

The build failure is
```
C/C++: [4/4] Linking CXX shared library app\lib\Debug\arm64-v8a\libvulkan_samples.so
C/C++: FAILED: app/lib/Debug/arm64-v8a/libvulkan_samples.so 
C/C++: cmd.exe /C "cd . && C:\Users\zieng\AppData\Local\Android\Sdk\ndk\25.2.9519653\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe --target=aarch64-none-linux-android30 --sysroot=C:/Users/zieng/AppData/Local/Android/Sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/windows-x86_64/sysroot -fPIC -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security   -DDEBUG=0 -fno-limit-debug-info  -static-libstdc++ -Wl,--build-id=sha1 -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments -shared -Wl,-soname,libvulkan_samples.so -o app\lib\Debug\arm64-v8a\libvulkan_samples.so @CMakeFiles\vulkan_samples.rsp  && cd ."
C/C++: ld: error: C:/Users/zieng/AppData/Local/Android/Sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/windows-x86_64/sysroot/lib/aarch64-linux-android/libc.a(memmove.o):(function memmove: .text+0x10): relocation R_AARCH64_CONDBR19 out of range: 1256256 is not in [-1048576, 1048575]; references __memcpy
C/C++: >>> referenced by memmove.S:95 (bionic/libc/arch-arm64/generic/bionic/memmove.S:95)
C/C++: >>> defined in C:/Users/zieng/AppData/Local/Android/Sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/windows-x86_64/sysroot/lib/aarch64-linux-android/libc.a(memcpy.o)
C/C++: ld: error: C:/Users/zieng/AppData/Local/Android/Sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/windows-x86_64/sysroot/lib/aarch64-linux-android/libc.a(wmemmove.o):(function wmemmove: .text+0x14): relocation R_AARCH64_CONDBR19 out of range: 1255180 is not in [-1048576, 1048575]; references __memcpy
C/C++: >>> referenced by memmove.S:95 (bionic/libc/arch-arm64/generic/bionic/memmove.S:95)
C/C++: >>> defined in C:/Users/zieng/AppData/Local/Android/Sdk/ndk/25.2.9519653/toolchains/llvm/prebuilt/windows-x86_64/sysroot/lib/aarch64-linux-android/libc.a(memcpy.o)
C/C++: clang++: error: linker command failed with exit code 1 (use -v to see invocation)
C/C++: ninja: build stopped: subcommand failed.
```

The full build log can be found at:
[build.log](https://github.com/user-attachments/files/18051763/build.log)

### PR DESCRIPTION
**Pull request by SaschaWillems**
## Description

The Android build setup for our samples is heavily outdated, making it impossible to locally build them with recent Android build tools (e.g. Android Studio) unless you find a way to manually download and setup the exact versions of the build related tools we require,

This makes it nigh impossible to easily build our samples on Android.

This PR fully updates our Android build scripts and setups to support up-to-date Android build tools, including the current version of Android Studio ("Ladybug"). So people wanting to build on Android can simply download Android studio, generate the project files and then easily build the samples.

It also fixes a few issue with our build files that only showed up with more recent build tools. One such change is a deprecation in the native bridge from C++ to Android, which required a fix in the framework (otherwise it would no longer compile).

**Note that this is a fairly pervasive change**. I did multiple builds, both release and debug from a clean setup to make sure this properly works. I also had to update some third party dependencies to make everything compile with updated compiler toolchains used on Android. I also updated the Android validation layers to the most recent release.

This hopefully fixes all the reported issues people had with trying to build our samples on Android.

Fixes #1243
Fixes #1188
Fixes #1185
Fixes #1178
Fixes #912

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [] I did a full batch run using the `batch` command line argument to make sure all samples still work properly
 Note: It does a famework change for Android, but afaik we don't have batch mode on Android. I did test several samples though.

### CODE REVIEW

File: docs/build.adoc
Comment (gpx1000) [inline docs/build.adoc]: Do we want to lock it down to Ladybug only?  Might be better to simply revert that back to highly recommended to install...  Also, the samples, don't strictly require Android Studio.  I can build / install / run them with just the tools; actually, that's what CI is doing come to think of it.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (gpx1000) [inline bldsys/cmake/template/gradle/app.build.gradle.in]: compileSdk should be bumped to 35

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (gpx1000) [inline bldsys/cmake/template/gradle/app.build.gradle.in]: Should also bump to the latest NDK while we're at it.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (gpx1000) [inline bldsys/cmake/template/gradle/app.build.gradle.in]: 35.0.0 is available.  Also might not need to use buildToolsVersion, actually I think it's deprecated?

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (gpx1000) [inline bldsys/cmake/template/gradle/app.build.gradle.in]: targetSdk should be set to 35

Comment (gpx1000) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: Github won't let me comment where it should be updated; but we can probably bump JAVA_VERSION from 1.8 up to VERSION_26

File: bldsys/cmake/template/gradle/download_vvl.gradle
Comment (gpx1000) [inline bldsys/cmake/template/gradle/download_vvl.gradle:79]: This is a note to myself to rethink how this is working.  Task dependencies in gradle don't need to be realized completely anymore in order to establish dependencies.  This closure forces gradle to completely define each task and then search for the same name.  This will slow the build down.

**Review by gpx1000 — CHANGES_REQUESTED** [review]

File: docs/build.adoc
Comment (SaschaWillems) [reply docs/build.adoc]: Yes, LadyBug is currently the only supported Android Studio Version, all others are EOL. Building from CLI still works fine, but most people (see issues) use Android Studio anyway. Aside from that even if you just want to build from CLI, having Android Studio pull all required components for you is much more convenient that trying to manually set them up ;)

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (SaschaWillems) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: I had most of those version updates (build tools, NDK) done by Android Studio's assistant. Getting this up and running took me roughly a month so I'd rather not want to change up things again.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (SaschaWillems) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: See comment above.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (SaschaWillems) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: See comment above.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (SaschaWillems) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: See comment above.

File: bldsys/cmake/template/gradle/download_vvl.gradle
Comment (SaschaWillems) [reply bldsys/cmake/template/gradle/download_vvl.gradle:79]: Yeah, we may look at this at a later point. It took me quite some time to figure out why this would fail until I realized we only hat mergeRelease... in there and I was trying to do a debug build :/

File: docs/build.adoc
Comment (gpx1000) [reply docs/build.adoc]: fair enough, just saying "required" is a strong word ;-)

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (gpx1000) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: I'm sorry Sascha, I meant to offer to help.  I went ahead and made a PR to your branch; feel free to accept / test out the changes to get the updates I'm suggesting.  There was one in there that is only deprecated, which means it'll disappear in a future Android Studio release.  might as well solve it now no?

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (gpx1000) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: Without having to muck about with gradle's wrapper, I bumped my suggested change to VERSION_21.  That should last us another 5 years ;-)

File: bldsys/cmake/template/gradle/download_vvl.gradle
Comment (gpx1000) [reply bldsys/cmake/template/gradle/download_vvl.gradle:79]: I hear yea, I'm not a fan of gradle's complicated way of task dependency.  Definitely NOT a today change.

**Comment by gpx1000** [comment]
I'll switch from Changes Requested to approved and leave it up to you on if you'd like to use the PR I made you.

**Review by gpx1000 — DISMISSED** [review]

**Comment by SaschaWillems** [comment]
Awesome. For whatever reason github didn't sent me a notification for that PR, so I didn't realize you already did all those changes :)

Will take a look at it tomorrow and merge. Very much appreciated 👍🏻

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (SaschaWillems) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: No offense taken :)

Merged your PR, so this should be addresed.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (SaschaWillems) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: Merged your PR, so this should be addressed.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (SaschaWillems) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: Merged your PR, so this should be addressed.

File: bldsys/cmake/template/gradle/app.build.gradle.in
Comment (SaschaWillems) [reply bldsys/cmake/template/gradle/app.build.gradle.in]: Merged your PR, so this should be addressed.

File: docs/build.adoc
Comment (SaschaWillems) [reply docs/build.adoc]: Totally agree, will adjust wording there to not confuse people :)

**Comment by SaschaWillems** [comment]
@gpx1000 The builds now sadly fail. It looks like NDK 28 and newer required the user to accept a license:

```
A problem occurred configuring project ':app'.
> com.android.builder.sdk.LicenceNotAcceptedException: Failed to install the following Android SDK packages as some licences have not been accepted.
     ndk;28.0.12433566 NDK (Side by side) 28.0.12433566
  To build this project, accept the SDK license agreements and install the missing components using the Android Studio SDK Manager.
  All licenses can be accepted using the sdkmanager command line tool:
  sdkmanager.bat --licenses
```

Running a .bat in a linux based CI obviously doesn't work here. Any ideas?

**Comment by gpx1000** [comment]
sdkmanager is an application that we use to download the tooling.  In windows it has a .bat in linux it's just a file.  I switched us away from using the docker image to using the preinstalled image which already has the NDK setup.  https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2004-Readme.md
NDK	26.3.11579264
27.2.12479018 (default)
That looks like the reason for the bug.  I guess for now, I'll write a bit more into the build for the CI to grab the 28 NDK version?  Or simply we can downgrade since the 28 version is technically in beta?  I'll leave the decision up to you Sascha.

**Comment by SaschaWillems** [comment]
I'm fine with both :)

In perspective going with NDK 28 is properly a bit more future proof.

**Comment by gpx1000** [comment]
I gave you a new PR that simply changes it to 27.  28 is still RC1 and in the future, it will still be asking for RC1 even after 28 is released if we left it the way it is.  So probably better to just use 27 for now.

**Comment by SaschaWillems** [comment]
Awesome 👍🏻

**Review by gpx1000 — DISMISSED** [review]

File: docs/build.adoc
Comment (SaschaWillems) [reply docs/build.adoc]: Changed wording a bit to make it clear that Ladybug is a requirement only when building directly through Android Studio. Hope that's better :)

Comment (gpx1000) [reply docs/build.adoc]: Perfect thanks

**Review by gpx1000 — APPROVED** [review]

**Review by gary-sweet — APPROVED** [review]
I can't check the Android specific stuff, but this doesn't hurt my build.

**Review by RodrigoHolztrattner-QuIC — APPROVED** [review]
Tested on Android (both Android Studio and gradle build flows) and it all looks good to me, samples are working, and I don't see any regressions.

**Comment by marty-johnson59** [comment]
3 approvals - merging

**Pull request merged**
